metric_learn/oasis.py

```diff
@@ -0,0 +1,293 @@
+    custom_order : User's custom order of triplets to feed oasis.
+    """
+    # Currently prepare_inputs makes triplets contain points and not indices
+    triplets = self._prepare_inputs(triplets, type_of_inputs='tuples')
+
+    # TODO: (Same as SCML)
+    # This algorithm is built to work with indices, but in order to be
+    # compliant with the current handling of inputs it is converted
+    # back to indices by the following fusnction. This should be improved
+    # in the future.
+    # Output: indices_to_X, X = unique(triplets)
+    triplets, X = self._to_index_points(triplets)
+
+    self.d = X.shape[1]  # (n_triplets, d)
+    self.n_triplets = triplets.shape[0]  # (n_triplets, 3)
+
+    self.components_ = self._check_M(self.custom_M)  # W matrix, needs self.d
+    if self.n_iter is None:
+      self.n_iter = self.n_triplets
+
+    # Get the order in wich the algoritm will be fed
+    if self.custom_order is not None:
+      self.indices = self._check_custom_order(self.custom_order)
+    else:
+      self.indices = self._get_random_indices(self.n_triplets,
+                                              self.n_iter,
+                                              self.shuffle,
+                                              self.random_sampling)
+
+    i = 0
+    while i < self.n_iter:
+        current_triplet = X[triplets[self.indices[i]]]
+        loss = self._loss(current_triplet)
+        vi = self._vi_matrix(current_triplet)
+        fs = self._frobenius_squared(vi)
+        # Global GD or Adjust to tuple
+        tau_i = np.minimum(self.c, loss / fs)
+
+        # Update components
+        self.components_ = np.add(self.components_, tau_i * vi)
+        i = i + 1
+
+    return self
+
+  def partial_fit(self, new_triplets, n_iter, shuffle=True,
+                  random_sampling=False, custom_order=None):
+    """
+    Reuse previous fit, and feed the algorithm with new triplets. Shuffle,
+    random sampling and custom_order options are available.
+
+    A new n_iter param can be set for the new_triplets.
+    """
+    self.n_iter = n_iter
+    self.fit(new_triplets, shuffle=shuffle, random_sampling=random_sampling,
+             custom_order=custom_order)
+
+  def _frobenius_squared(self, v):
+    """
+    Returns Frobenius norm of a point, squared
+    """
+    return np.trace(np.dot(v, v.T))
+
+  def _loss(self, triplet):
+    """
+    Loss function in a triplet
+    """
+    S = -1 * self.score_pairs([[triplet[0], triplet[1]],
+                              [triplet[0], triplet[2]]])
+    return np.maximum(0, 1 - S[0] + S[1])
+
+  def _vi_matrix(self, triplet):
+    """
+    Computes V_i, the gradient matrix in a triplet
+    """
+    # (pi+ - pi-)
+    diff = np.subtract(triplet[1], triplet[2])  # Shape (, d)
+    result = []
+
+    # For each scalar in first triplet, multiply by the diff of pi+ and pi-
+    for v in triplet[0]:
+        result.append(v * diff)
+
+    return np.array(result)  # Shape (d, d)
+
```

Comment:
@bellet So this is the core part of OASIS, I've separated `_loss`, `_frobenius_squared` and `_vi_matrix` on purpose to make the code more readable.

Small comments:

`self.indices` has the order in which the triplets will be seen
`self.n_iter` matches the size of self.indices. It's the number of iterations
`self.components_` is in this case the whole M matrix of bilinear similarity, as there is no transformation here
`partial_fit` does the same as fit but with new triplets and reusing the existing `self.components_` in the object

Paper is [here ](https://www.jmlr.org/papers/volume11/chechik10a/chechik10a.pdf)

Reply:
Another note: In line 90, `loss / fs` can give a warning about zero division, when `fs` is zero, because `_vi` is zero. This happens particularly when a triplet with zeros in all features is seen.

In that case `loss/fs` will become `+inf`, but `np.minimum` will always choose `self.c`